New CNN model

jamesrequa opened the pull request:
- Implemented new CNN architecture
- Added BatchNormalization
- Added learning rate as a parameter in train_model method of
keras_helper and added as a user defined hyper-parameter in the
notebook
- Added methods for save_weights and load_weights to keras_helper;
functionality also working in the notebook
- Moved epochs to training block of notebook for easier access
- Added optional Continue Training block to notebook
- Fixed bugs with EarlyStopping and ModelCheckpoint

jamesrequa committed:
Merge branch 'master' into new-CNN-model

jamesrequa committed:
Merge branch 'master' into new-CNN-model

jamesrequa committed:
New CNN Architecture

- Implemented new CNN architecture
- Added BatchNormalization
- Added learning rate as a parameter in train_model method of
keras_helper and added as a user defined hyper-parameter in the
notebook
- Added methods for save_weights and load_weights to keras_helper;
functionality also working in the notebook
- Moved epochs to training block of notebook for easier access
- Added optional Continue Training block to notebook
- Fixed bugs with EarlyStopping and ModelCheckpoint

godardt committed:
Merge with master

godardt committed:
Remove ipynb notebook

EKami commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -223,17 +224,36 @@
+
+epochs = 5
+learn_rate = 0.001
+train_losses, val_losses, fbeta_score = classifier.train_model(x_train, y_train, learn_rate, epochs, batch_size, 
```
Here you are overriding the `train_losses` and `val_losses` arrays from the 1st training pass. Instead you should append the results to the previous ones or do 2 separate plots below

godardt commented on src/keras_helper.py:
```diff
@@ -31,51 +31,71 @@ def __init__(self):
+        self.classifier.add(Conv2D(32, (3, 3), padding='same', activation='relu'))
         self.classifier.add(Conv2D(32, (3, 3), activation='relu'))
-        self.classifier.add(MaxPooling2D(pool_size=(2, 2)))
+        self.classifier.add(MaxPooling2D(pool_size=2))
```
This is actually the same thing but it's just a matter of preferences. I wanted to keep a tuple to not forget that it's the 2x2 patch but you can keep it like that if you want :)

jamesrequa replied:
Yea I prefer just having one integer since it defaults to equal pool dims, hope you don't mind! ha

godardt replied:
Yes np :)

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -223,17 +224,36 @@
+
+epochs = 5
+learn_rate = 0.001
+train_losses, val_losses, fbeta_score = classifier.train_model(x_train, y_train, learn_rate, epochs, batch_size, 
```
Here you are overriding the `train_losses` and `val_losses` arrays from the 1st training pass. Instead you should append the results to the previous ones or do 2 separate plots below

jamesrequa replied:
Thanks for catching this. Revising to append results  for `train_losses` and `val_losses` each time so its not overwritten.

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -208,7 +209,7 @@
+from tensorflow.contrib.keras.api.keras.callbacks import ModelCheckpoint
 
 filepath="weights.best.hdf5"
 checkpoint = ModelCheckpoint(filepath, monitor='val_acc', verbose=1, save_best_only=True, mode='max')
```
The `mode='max'` is not needed here. The default value is ok for our case :)

jamesrequa replied:
OK will remove it

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -208,7 +209,7 @@
+from tensorflow.contrib.keras.api.keras.callbacks import ModelCheckpoint
 
 filepath="weights.best.hdf5"
 checkpoint = ModelCheckpoint(filepath, monitor='val_acc', verbose=1, save_best_only=True, mode='max')
```
Are you sure we want to keep the `verbose=1`? It really generates ugly output and doesn't give us much useful information:
<img width="1005" alt="amazon_forest_notebook" src="https://cloud.githubusercontent.com/assets/4030626/26435667/de93a918-40de-11e7-8f02-dcf3c7d9d83f.png">

Something like this would be best:
`checkpoint = ModelCheckpoint(filepath, monitor='val_acc', save_best_only=True)`
Wdyt?

jamesrequa replied:
Yes the output isn't pretty but its nice to see at a quick glance that the model has achieved a new best weight on a particular epoch. Otherwise you have to search for the accuracy numbers to see for yourself which one is the best.

godardt replied:
Ok we'll keep it verbose then :)

godardt reviewed (changes requested):
Great job!

jamesrequa committed:
New CNN model #8 - Revisions

- Removed mode='max' from checkpoint
- Added a learning rate optimization schedule which in the process also
resolves the train_losses and val_losses being overwritten

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -212,39 +210,40 @@
-# ## Continue Training
+train_losses, val_losses = [], []
+epochs_arr = [20, 5, 5]
+learn_rates = [0.001, 0.0001, 0.00001]
```
We could move these parameters (epoch and learn_rates) in the `Define hyperparameters` cell or change/remove the `Define hyperparameters` cell. It doesn't make much sense to have hyperparameters spread in the notebook when there is 1 place where they should be defined :)

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -212,39 +210,40 @@
+
+# Here we define the model and begin training. 
+# 
+# Before starting the training process, you should first set a learning rate annealing optimization schedule by choosing a series of learning rates (learn_rates) with corresponding number of epochs for each (epochs_arr).
```
In markdown code variables are usually contained in ` quotes.

...schedule by choosing a series of learning rates (`learn_rates`) with...

...number of epochs for each (`epochs_arr`).

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -208,32 +207,52 @@
+                                                                           batch_size, validation_split_size=validation_split_size, 
+                                                                           train_callbacks=[checkpoint])
+    train_losses.append(tmp_train_losses)
+    val_losses.append(tmp_val_losses)
```
This will append lists into lists, instead we want to append the value to the end of the original list so the correct behaviour here would be using `+=` instead of `append`

jamesrequa committed:
Revisions - PR #8

- Fixed issue with appending train_losses and val_losses
- Moved all hyperparameters into one code block prior to training for
easy reference
- Image resize is now its own block as part of the preprocessing steps

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -214,15 +210,32 @@
+# 
+# If you'd like to create a learning rate annealing schedule you can set a series of learning rates with `lr_1, lr_2, lr_3` and corresponding number of epochs for each `epochs_1, epochs_2, epochs_3`. 
+# 
+# Alternatively, if you just want to run one training session at a fixed learning rate and num epochs you can just enter values for `lr_1` and `epochs_2` only.
```
Following these instructions I'll be tempted to define only the following learning rates and epochs:
```
lr_1 = 0.001
epochs_1 = 10
```
Except that if I do that the code below will crash :P

It's better to use directly the epoch and lr arrays and let the user add as much values as we want in it :)

jamesrequa replied:
Oops I should have tested it first to see if that would work. OK.

jamesrequa replied:
@EKami Changes already made, please update review!

jamesrequa committed:
PR #8 - Markdown edits

- Clarifications made to the hyperparameters markdown with regards to
learning rate annealing.

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -201,39 +198,72 @@
+# 
+# Choose your hyperparameters below for training. 
+# 
+# We have assigned a series of learning rate values to create an optimized annealing schedule `lr_1, lr_2, lr_3` and corresponding number of epochs for each `epochs_1, epochs_2, epochs_3`. Feel free to change these values or you can just use the defaults we have filled in. 
```
Hmm, sorry for not being clear but what I wanted to say is that defining variables which are just used to fill an array just adds a burden to the code 😅 .  For instance if I want to add another value to the array and follow this pattern I would have to add `lr_4` and `epochs_4` then add them to the array (where I could just add them to the array). It would be much easier to directly have the arrays (such as `epochs_arr = [0.001, 0.0001, 0.00001]`) and explain in the comment why we are creating such arrays. These kind of behaviours can often lead to what we call [edge cases](https://en.wikipedia.org/wiki/Edge_case) in the dev world.

jamesrequa committed:
PR#8 Revisions - Markdown edits #2

- Simplified hyperparameter markdown

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -213,27 +213,19 @@
+# 
+# Here we define the model and begin training. 
+# 
+# Note that we have created a learning rate annealing schedule with a series of learning rates as defined `lr_1, lr_2, lr_3` and corresponding number of epochs for each `epochs_1, epochs_2, epochs_3`. Feel free to change these values if you like or just use the defaults. 
```
> we have created a learning rate annealing schedule with a series of learning rates as defined `lr_1, lr_2, lr_3` and corresponding number of epochs for each `epochs_1, epochs_2, epochs_3`

This message is outdated x)

jamesrequa replied:
@EKami Sorry about that...fixed

jamesrequa committed:
PR#8 Revisions - markdown for lr annealing

- Fixed reference to variables that were already removed

jamesrequa commented:
![](https://github.trello.services/images/mini-trello-icon.png) [BatchNorm](https://trello.com/c/6yxxWsW7/2-batchnorm)

jamesrequa commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Learning Rate Annealing/Simulated annealing](https://trello.com/c/Hjs2DbJV/11-learning-rate-annealing-simulated-annealing)
